### ISSUE TITLE
Import not thread-save?

### ISSUE THREAD

**Issue by htgoebel** [pyinstaller/pyinstaller#2371]
The test-case [test_threading_module](https://github.com/htgoebel/pyinstaller/blob/c778285/tests/functional/test_basic.py#L425) fails occasionally for Python 3.4 and 3.5. When re-running the test-job, it sometimes passes or fails again.

Example of failing job: https://travis-ci.org/htgoebel/pyinstaller/jobs/190919975, while [this one](https://travis-ci.org/htgoebel/pyinstaller/jobs/190904243) passed when re-running.

This test fails with
```
AttributeError: module 'pyi_testmod_threading' has no attribute 'x'
```
in one of the threads.

This looks like the module gets added to the list of modules in the first thread but it's body is not yet executed. Then thread 2 comes along, finds the module in the list, assumes it is imported and tries to access `x`. But since the body of the module is not yet executed, this attribute does not yet exist.

[pyimod03_importers.py](https://github.com/htgoebel/pyinstaller/blob/c778285/PyInstaller/loader/pyimod03_importers.py#L30) reads:

> In Python 3.3+ tne locking scheme has changed to per-module locks for the most part.
> Global locking should not be required in Python 3.3+

Which matches the analysis may be the cause of the problem.

**Comment by htgoebel** [comment]
Adding a delay to `pyi_testmod_threading` to trigger the timing problem dis not make the test-case more reliable :-(

**Comment by xoviat** [comment]
The [builtin zip importer][1] seems to be [calling][2] [`exec_module`][3] whereas PyInstaller is [calling][4] [`load_module`][5]. It almost seems like it would be easier to inherit from the builtin zip importer and then override the [`get_data`][6] method.

Also, the documentation says the following about `load_module`:

> If the requested module already exists in sys.modules, that module should be used **and reloaded**. Otherwise the loader should create a new module and insert it into sys.modules before any loading begins, to prevent recursion from the import. If the loader inserted a module and the load fails, it must be removed by the loader from sys.modules; modules already in sys.modules before the loader began execution should be left alone (see importlib.util.module_for_loader()).

It appears that `load_module` [has been deprecated] in favor of `exec_module`, and although this hasn't yet been changed in the builtin zip importer class, it probably will be.

[1]: https://github.com/python/cpython/blob/f344c77676450ece100b03165f1ec542c34286fe/Modules/zipimport.c#L1129
[2]: https://github.com/python/cpython/blob/f344c77676450ece100b03165f1ec542c34286fe/Modules/zipimport.c#L510
[3]: https://docs.python.org/3/library/importlib.html#importlib.abc.Loader.exec_module
[4]: https://github.com/htgoebel/pyinstaller/blob/c778285/PyInstaller/loader/pyimod03_importers.py#L573
[5]: https://docs.python.org/3/library/importlib.html#importlib.abc.Loader.load_module
[6]: https://github.com/python/cpython/blob/f344c77676450ece100b03165f1ec542c34286fe/Modules/zipimport.c#L1129
[7]: https://www.python.org/dev/peps/pep-0451/#other-api-additions

**Comment by xoviat** [comment]
While probing this further, I received the following interesting traceback:

```
t1 started
t2 started
main started
t2 <module 'pyi_testmod_threading' (<zipimporter object "???">)>
main <module 'pyi_testmod_threading' (<zipimporter object "???">)>
Traceback (most recent call last):
  File "/tmp/pytest-of-travis/pytest-0/popen-gw0/test_threading_module_onefile_0/test_threading_module.py", line 22, in <module>
Exception in thread Thread-2:
Traceback (most recent call last):
  File "/opt/python/3.5.2/lib/python3.5/threading.py", line 914, in _bootstrap_inner
    self.run()
  File "/opt/python/3.5.2/lib/python3.5/threading.py", line 862, in run
    self._target(*self._args, **self._kwargs)
  File "/tmp/pytest-of-travis/pytest-0/popen-gw0/test_threading_module_onefile_0/test_threading_module.py", line 14, in doit
    print_(nm, pyi_testmod_threading.x)
AttributeError: module 'pyi_testmod_threading' has no attribute 'x'
    doit('main')
  File "/tmp/pytest-of-travis/pytest-0/popen-gw0/test_threading_module_onefile_0/test_threading_module.py", line 14, in doit
    print_(nm, pyi_testmod_threading.x)
AttributeError: module 'pyi_testmod_threading' has no attribute 'x'
t1 5
t1 <module 'pyi_testmod_threading' (<zipimporter object "???">)>
```

It looks like `pyi_testmod_threading` is actually the loader object.

**Comment by htgoebel** [comment]
@xoviat Looks like you have fun digging into this kind of problem. Nice to have a companion for this :-)

I'm quite sure `pyi_testmod_threading` actually is a module, as the output says: `<module 'pyi_testmod_threading' …>`. The `(<zipimporter object "???">)` is just the path.

Regarding changing to the builtin zip importer and exec_module: This may be an option, but may also need larger changes to the import mechanism. I'm currently short in time, but if you'd like to look at this, it would be find. I suggest working on this on a separate branch.

**Comment by xoviat** [comment]
@htgoebel This and py36 are completed. py36 resolves the "yield" syntax error that is currently failing the tests.

**Comment by xoviat** [comment]
@htgoebel This should be closed.

**Comment by htgoebel** [comment]
Closed by 37b4046af5d44a1687041658c989dcea78879d02, see #2377.
@xoviat Thanks for pointing to this :-)

**Comment by ThomasWaldmann** [comment]
@htgoebel the last few comments are confusing. so, what is the state of this issue?

you say "closed by", but it is still open.

**Comment by xoviat** [comment]
This is resolved for non c-extensions on Python >= 3.4.

**Comment by htgoebel** [comment]
@TW Oh, looks like I simply forgot to close this issue.
